drivers/ethernet/phy/phy_tja1103.c

```diff
@@ -0,0 +1,286 @@
+#define TJA1103_DEVICE_CONTROL_GLOBAL_CFG_EN BIT(14)
+#define TJA1103_DEVICE_CONTROL_SUPER_CFG_EN  BIT(13)
+/* Shared - PHY control register */
+#define TJA1103_PHY_CONTROL                  (0x8100U)
+#define TJA1103_PHY_CONTROL_CFG_EN           BIT(14)
+/* Shared - PHY status register */
+#define TJA1103_PHY_STATUS                   (0x8102U)
+#define TJA1103_PHY_STATUS_LINK_STAT         BIT(2)
+
+struct phy_tja1103_config {
+	const struct device *mdio;
+	uint8_t phy_addr;
+	uint8_t master_slave;
+};
+
+struct phy_tja1103_data {
+	struct phy_link_state state;
+	struct k_sem sem;
+};
+
+static inline int phy_tja1103_c22_read(const struct device *dev, uint16_t reg, uint16_t *val)
+{
+	const struct phy_tja1103_config *const cfg = dev->config;
+
+	return mdio_read(cfg->mdio, cfg->phy_addr, reg, val);
+}
+
+static inline int phy_tja1103_c22_write(const struct device *dev, uint16_t reg, uint16_t val)
+{
+	const struct phy_tja1103_config *const cfg = dev->config;
+
+	return mdio_write(cfg->mdio, cfg->phy_addr, reg, val);
+}
+
+static inline int phy_tja1103_c45_write(const struct device *dev, uint16_t devad, uint16_t reg,
+					uint16_t val)
+{
+	const struct phy_tja1103_config *cfg = dev->config;
+
+	return mdio_write_c45(cfg->mdio, cfg->phy_addr, devad, reg, val);
+}
+
+static inline int phy_tja1103_c45_read(const struct device *dev, uint16_t devad, uint16_t reg,
+				       uint16_t *val)
+{
+	const struct phy_tja1103_config *cfg = dev->config;
+
+	return mdio_read_c45(cfg->mdio, cfg->phy_addr, devad, reg, val);
+}
+
+static int phy_tja1103_reg_read(const struct device *dev, uint16_t reg_addr, uint32_t *data)
+{
+	const struct phy_tja1103_config *cfg = dev->config;
+	int ret;
+
+	mdio_bus_enable(cfg->mdio);
+
+	ret = phy_tja1103_c22_read(dev, reg_addr, (uint16_t *)data);
+
+	mdio_bus_disable(cfg->mdio);
+
+	return ret;
+}
+
+static int phy_tja1103_reg_write(const struct device *dev, uint16_t reg_addr, uint32_t data)
+{
+	const struct phy_tja1103_config *cfg = dev->config;
+	int ret;
+
+	mdio_bus_enable(cfg->mdio);
+
+	ret = phy_tja1103_c22_write(dev, reg_addr, (uint16_t)data);
+
+	mdio_bus_disable(cfg->mdio);
+
+	return ret;
+}
+
+static int phy_tja1103_id(const struct device *dev, uint32_t *phy_id)
+{
+	uint16_t val;
+
+	if (phy_tja1103_c22_read(dev, MII_PHYID1R, &val) < 0) {
+		return -EIO;
+	}
+
+	*phy_id = (val & UINT16_MAX) << 16;
+
+	if (phy_tja1103_c22_read(dev, MII_PHYID2R, &val) < 0) {
+		return -EIO;
+	}
+
+	*phy_id |= (val & UINT16_MAX);
+
+	return 0;
+}
+
+static int phy_tja1103_await_phyid(const struct device *dev)
+{
+	int ret;
+	uint32_t count;
+	uint32_t phy_id;
+
+	k_sleep(K_USEC(100000));
+
+	/**
+	 * wait until out of reset.
+	 */
+	for (count = 0U; count < TJA1103_AWAIT_RETRY_COUNT; ++count) {
+		ret = phy_tja1103_id(dev, &phy_id);
+		if (ret >= 0) {
+			if (phy_id == TJA1103_ID) {
+				break;
+			}
+			ret = -ETIMEDOUT;
+		}
+		k_sleep(K_USEC(TJA1103_AWAIT_DELAY_POLL_US));
+	}
```

Comment:
I wonder if lot of this boilerplate code could be replaced by `WAIT_FOR()` macro?